libass/ass_filesystem.c

```diff
@@ -50,6 +50,41 @@ static bool alloc_path(ASS_Dir *dir, size_t size)
 }
 
 
+#if HAVE_FSTAT || defined(_WIN32)
+// Microsoft docs require types.h to be included before stat.h
+// In POSIX >= Issue 6 just stat.h suffices, but include both anyway
+#include <sys/types.h>
+#include <sys/stat.h>
+
+#if defined(_WIN32)
+#define fstat(...) _fstati64(__VA_ARGS__)
+#define stat _stati64
+#ifndef S_ISREG
+// MinGW already defines this but pure Microsoft docs don't mention it
+#define S_ISREG(m) ((m & _S_IFMT) == _S_IFREG)
+#endif
+#endif
+
+bool ass_get_file_size(const FILE *fp, size_t *size)
+{
+    struct stat info;
+    int fd = fileno((FILE *) fp);
+    if (fstat(fd, &info))
+        return false;
+
+    if (!S_ISREG(info.st_mode) || info.st_size > SIZE_MAX)
```

Comment:
For > 4G files at 32-bit system that triggers failure, so `ass_load_file` fallbacks to incremental read which eventually fail due to out of memory. I think `ass_get_file_size` should return `uint64_t` (or `int64_t` which can be combined with &minus;1 error) and let caller handle overflow.

Reply:
I think a function with the name `ass_get_file_size` should return a known file size even if that size is larger than `SIZE_MAX`. Currently we don't have a use for such sizes, but that can change in the future (for example, for a detailed error message).

Reply:
While we are presumably unlikely to encounter files larger than >2^64 bytes soon, POSIX’s `stat` struct records the size in `off_t` (Windows’s `fstati64` as a 64bit type) and `off_t` might be *any* width, so the return type would need to be `intmax_t`.
However we only load two types of files:
 - font files, which we’ll need to fully load into memory and thus limit to `SIZE_MAX` anyway
  - ASS styles or full files for which we want to switch to a streaming approach so `get_file_size` will be irrelevant *(i was intending to do that some later date, but if optimising for this case becomes a problem perhaps I should already do that now)*

I’d be surprised if we ever have to deal with more than those types and therefore don't see where `get_file_size > SIZE_MAX` would ever be useful. Do you have some future usecase in mind?

Reply:
I feel it's incorrect to guess intended use of the result inside a function, it should be caller responsibility. Especially if there's already `bufmax > SIZE_MAX - 2` check outside of it. As for > 2^64 (or 2^63) file sizes, we can safely treat it as error for any foreseeable future. And the simplest use of large result is error reporting: "Cannot load file '...' as it's too large (5.15GB)".

Reply:
Isn’t that a contradiction? Treating sizes above 2^64 as an error is exactly what Oneric is doing here, and (from what I understand) it’s impossible to move that handling into the caller except by changing the return type either to `intmax_t` or to a platform-dependent type (or typedeffing `off_t` on Windows).

If the issue is just the fact that the size is being returned in a parameter rather than the return value, how about returning `size_t`, using `SIZE_MAX` as the error marker, and treating files of size exactly `SIZE_MAX` as an error case as well?

Reply:
Or you could also use 0 as the error marker.

Reply:
Ah, I forgot we need two error modes. So use one of those for one mode and the other for the other.

Reply:
We want:
 - ideally quit early on empty files
 - definitely quit early on files `>= SIZE_MAX-1` (offset by 1 because as long as this code also handles ASS files we need an additional terminating zero)
 - dynamically increase the buffer for unknown size
 - use an exactly fitting buffer for known size
 
 With `SIZE_MAX` and `0` as error values we can’t handle all of them iinm, but if we sacrifice the early exit on empty files it could work. Though I must say I’m not a fan of using a in-band value which could technically be valid as an error code, I guess it works here (too bad we don’t have `ssize_t` on Win32).
 
 But if the issue is just with the name not reflecting the `size_t` restriction, how about renaming it to `ass_get_file_buffer_size`?

Reply:
> too bad we don’t have `ssize_t` on Win32

That would merely _halve_ the valid value space, as opposed to reserving two individual values.

I’m not sure how it isn’t enough, though: you want to handle empty files and huge files the same. In fact, if the upper limit is `>= SIZE_MAX-1`, then you don’t even sacrifice any valid values at all and you even have an extra value left over!

* `0` normally means an empty file, and you want to discard empty files. Return `0` for huge files, too, and they’ll be discarded, too.
* Values from `1` to `SIZE_MAX-2` inclusive are valid sizes of files whose size is known.
* If the size is unknown, either `SIZE_MAX-1` or `SIZE_MAX` can in theory be used as a sentinel value, and the other remains unused.

Reply:
> `0` normally means an empty file, and you want to discard empty files. Return `0` for huge files, too, and they’ll be discarded, too.

Then we can't log a warning though for too large files (which MrSmile mentioned above and I’m adding locally right now)

Reply:
> > too bad we don’t have ssize_t on Win32
>
> That would merely halve the valid value space, as opposed to reserving two individual values.

hmm, I assumed `ssize_t` can hold at least `-SIZE_MAX` to `SIZE_MAX`, but apparently this is not guaranteed and POSIX allows for environments where `ssize_t` is only as wide as `long` (identifed by confstr’s `_CS_POSIX_V7_WIDTH_RESTRICTED_ENVS`).
This makes it less useful then, yeah.

Reply:
I mean, there’s not much choice on platforms with 64-bit `size_t`. (Some) common compilers only got 128-bit support relatively recently, if indeed at all.

> Then we can't log a warning though for too large files

Why not log it inside this function?

Reply:
I don't care about platforms with 64-bit `size_t`, such large files would be unfeasible for many years to come. I care about platforms with 32-bit `size_t`, but > 4G files. In such cases file size is perfectly known, but `ass_get_file_size` returns a error, which I think isn't right, especially if that function would be reused in another context in the future.

Reply:
> especially if that function would be reused in another context in the future.

At least with what I have in mind, it `ass_get_file_size` will only ever be used for files we fully load into memory (font files and atm ASS files), while future streamed ASS files won't use it at all (instead read fixed size chunks like the second commit of this series to have an almost fixed memory usage (if lines are bigger than CHUNK_SIZE realloc will still be needed))

Do you prefer one of the alternative names or have another name suggestion to reflect the limitation?

Reply:
It's worth noting that even if we stream a larger file in the future, we still load all events into memory, so unless the file is full of comment lines or embedded fonts, we'll likely run out of 32-bit memory space anyway.